### PR TITLE
Preliminary version of SearchResultsAggregator denoting feature milestone

### PR DESCRIPTION
**Pull request by alpaix**
Implemented a resource responsible for merging search results. 
The resource is not integrated to main UI search flows yet.
Next milestone will be around creating a `SourceRepository` handling all-sources scenario.

Changes:
da7bde5 - 5 days ago : Search results aggregator
de6bbaf - 2 weeks ago : Search Aggregator proxy service

### CODE REVIEW

**Comment by alpaix** [comment]
//cc @yishaigalatzer @emgarten

**Comment by alpaix** [comment]
@johnataylor I refactored your PoC and handled border cases. Please review.

File: SearchAggregator.sln [hunk 1]
```diff
@@ -0,0 +1,22 @@
+﻿
+Microsoft Visual Studio Solution File, Format Version 12.00
+# Visual Studio 14
+VisualStudioVersion = 14.0.24720.0
+MinimumVisualStudioVersion = 10.0.40219.1
+Project("{8BB2217D-0F2D-49D1-97BC-3654ED321F3B}") = "SearchAggregator", "src\SearchAggregator\SearchAggregator.xproj", "{1086DCD1-50C1-4A86-89EE-7ECB92C2B43F}"
```
Comment (alpaix) [inline SearchAggregator.sln]: SearchAggregator proxy service won't be merged into dev. It used to be a prototyping playground. Will submit a cleaner PR changes before merging.

**Comment by alpaix** [comment]
Let me know if you think this PR can be merged or should I keep it until the feature is ready.

File: src/NuGet.Core/NuGet.Indexing/project.json [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    "../../../submodules/NuGet.Services.Metadata/src/NuGet.Indexing/OwnerAnalyzer.cs",
+    "../../../submodules/NuGet.Services.Metadata/src/NuGet.Indexing/VersionAnalyzer.cs",
+    "../../../submodules/NuGet.Services.Metadata/src/NuGet.Indexing/CamelCaseFilter.cs",
+    "../../../submodules/NuGet.Services.Metadata/src/NuGet.Indexing/TagsAnalyzer.cs",
+    "../../../submodules/NuGet.Services.Metadata/src/NuGet.Indexing/SemanticVersionFilter.cs"
+  ],
+  "dependencies": {
+    "EntityFramework": "5.0.0",
+    "Lucene.Net": "3.0.3",
+    "Lucene.Net.Contrib": "3.0.3",
```
Comment (yishaigalatzer) [inline src/NuGet.Core/NuGet.Indexing/project.json:22]: We can't pull in lucene.net.contrib or lucene.net.store.azure, also not sharpziplib or storage into the client for tis

File: src/NuGet.Core/NuGet.Protocol.Core.Types/Model/PackageSearchMetadata.cs [hunk 3]
```diff
@@ -0,0 +1,27 @@
+﻿using NuGet.Packaging.Core;
+using System;
+using System.Collections.Generic;
+
+namespace NuGet.Protocol.Core.Types
+{
+    public class PackageSearchMetadata : IPackageSearchMetadata
```
Comment (yishaigalatzer) [inline src/NuGet.Core/NuGet.Protocol.Core.Types/Model/PackageSearchMetadata.cs:7]: Don't we already have an equivalent type for this?

File: src/NuGet.Core/NuGet.Protocol.VisualStudio/Converters/PackageSearchMetadataConverter.cs [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+
+        public override object ReadJson(JsonReader reader, Type objectType, object existingValue, JsonSerializer serializer)
+        {
+            if (reader.TokenType == JsonToken.Null)
+            {
+                return null;
+            }
+
+            // Load JObject from stream
+            JObject jObject = JObject.Load(reader);
```
Comment (yishaigalatzer) [inline src/NuGet.Core/NuGet.Protocol.VisualStudio/Converters/PackageSearchMetadataConverter.cs:28]: Is this the most efficient way to write a converter here?

File: src/NuGet.Core/NuGet.Protocol.Core.Types/Model/PackageSearchMetadata.cs (hunk 3 above)
Comment (alpaix) [inline src/NuGet.Core/NuGet.Protocol.Core.Types/Model/PackageSearchMetadata.cs:7]: Good point. Current type(s) are incomplete and contain wire protocol or UI specific data. Therefore I couldn't use them for merging as they are.
Planning to refactor 'em by next milestone into a unified class or family of classes.

File: src/NuGet.Core/NuGet.Protocol.VisualStudio/Resources/AggregatedSearchResource.cs [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+        private readonly ISearchResource[] _searchResources;
+
+        public AggregatedSearchResource(IEnumerable<ISearchResource> searchResources)
+        {
+            _searchResources = searchResources.ToArray();
+        }
+
+        public Task<IEnumerable<PackageSearchMetadata>> SearchAsync(string searchTerm, SearchFilter filters, int skip, int take, CancellationToken cancellationToken)
+        {
+            throw new NotImplementedException();
```
Comment (yishaigalatzer) [inline src/NuGet.Core/NuGet.Protocol.VisualStudio/Resources/AggregatedSearchResource.cs]: Don't merge unfinished code

File: src/NuGet.Core/NuGet.Protocol.VisualStudio/Resources/ISearchResultsIndexer.cs [hunk 6]
```diff
@@ -0,0 +1,10 @@
+﻿using NuGet.Protocol.Core.Types;
+using System.Collections.Generic;
+
+namespace NuGet.Protocol.VisualStudio
+{
+    public interface ISearchResultsIndexer : INuGetResource
+    {
+        IDictionary<string, int> Rank(string queryString, IEnumerable<PackageSearchMetadata> entries);
```
Comment (yishaigalatzer) [inline src/NuGet.Core/NuGet.Protocol.VisualStudio/Resources/ISearchResultsIndexer.cs]: This shouldn't be a reasource

File: src/NuGet.Core/NuGet.Protocol.VisualStudio/Resources/ISearchResultsAggregator.cs [hunk 7]
```diff
@@ -0,0 +1,10 @@
+﻿using NuGet.Protocol.Core.Types;
+using System.Collections.Generic;
+
+namespace NuGet.Protocol.VisualStudio
+{
+    public interface ISearchResultsAggregator : INuGetResource
+    {
+        IEnumerable<PackageSearchMetadata> Aggregate(string queryString, params IEnumerable<PackageSearchMetadata>[] results);
```
Comment (yishaigalatzer) [inline src/NuGet.Core/NuGet.Protocol.VisualStudio/Resources/ISearchResultsAggregator.cs]: This is not a resource

File: src/NuGet.Core/NuGet.Protocol.VisualStudio/Resources/LuceneSearchResultsIndexer.cs [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+using NuGet.Indexing;
+using NuGet.Protocol.Core.Types;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace NuGet.Protocol.VisualStudio
+{
+    public class LuceneSearchResultsIndexer : ISearchResultsIndexer
+    {
+        public LuceneSearchResultsIndexer()
```
Comment (yishaigalatzer) [inline src/NuGet.Core/NuGet.Protocol.VisualStudio/Resources/LuceneSearchResultsIndexer.cs]: no need for a default constructor

File: src/NuGet.Core/NuGet.Protocol.VisualStudio/project.json [hunk 9]
```diff
@@ -14,7 +14,9 @@
     "warningsAsErrors": true
   },
   "dependencies": {
+    "Lucene.Net": "3.0.3",
```
Comment (yishaigalatzer) [inline src/NuGet.Core/NuGet.Protocol.VisualStudio/project.json]: This gets pulled in by indexing

File: src/SearchAggregator/Controllers/PackagesController.cs [hunk 10]
```diff
@@ -0,0 +1,66 @@
+﻿using Microsoft.AspNet.Mvc;
```
Comment (yishaigalatzer) [inline src/SearchAggregator/Controllers/PackagesController.cs]: No MVC dependencies

File: src/SearchAggregator/Controllers/PackagesController.cs [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+using Microsoft.Extensions.OptionsModel;
+using Newtonsoft.Json.Linq;
+using SearchAggregator.Services;
+using System;
```
Comment (yishaigalatzer) [inline src/SearchAggregator/Controllers/PackagesController.cs]: Sort dependencies, system first

File: src/SearchAggregator/Controllers/PackagesController.cs [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+    {
+        private readonly ProxySettings _settings;
+
+        public PackagesController(IOptions<ProxySettings> settingsOptions)
+        {
+            _settings = settingsOptions.Value;
+        }
+
+        [HttpGet("/api/packages/search")]
+        public async Task<IActionResult> Search(IDictionary<string, string> queryParams)
```
Comment (yishaigalatzer) [inline src/SearchAggregator/Controllers/PackagesController.cs]: this is not how we are merging this into the product

File: src/NuGet.Core/NuGet.Indexing/project.json (hunk 2 above)
Comment (alpaix) [inline src/NuGet.Core/NuGet.Indexing/project.json:22]: Ananlyzers won't build without `Lucene.Net.Contrib`
@johnataylor Do you know if it's possible to get rid of it while maintaining correct indexing functionality?

File: src/NuGet.Core/NuGet.Protocol.VisualStudio/Resources/AggregatedSearchResource.cs (hunk 5 above)
Comment (alpaix) [inline src/NuGet.Core/NuGet.Protocol.VisualStudio/Resources/AggregatedSearchResource.cs]: Okay

File: src/SearchAggregator/Controllers/PackagesController.cs (hunk 10 above)
Comment (alpaix) [inline src/SearchAggregator/Controllers/PackagesController.cs]: As I mentioned earlier will remove proxy service from the final change.

Comment (yishaigalatzer) [inline src/SearchAggregator/Controllers/PackagesController.cs]: Don't want to merge unrelated code into dev. You can keep working in your branch

File: src/SearchAggregator/Controllers/PackagesController.cs (hunk 12 above)
Comment (alpaix) [inline src/SearchAggregator/Controllers/PackagesController.cs]: Removing along with the proxy service

File: src/NuGet.Core/NuGet.Protocol.VisualStudio/Resources/ISearchResultsIndexer.cs (hunk 6 above)
Comment (alpaix) [inline src/NuGet.Core/NuGet.Protocol.VisualStudio/Resources/ISearchResultsIndexer.cs]: This is arguable. But I'd prefer to postpone this discussion for now. Will make it a regular class and push as a functional component first.

**Comment by alpaix** [comment]
Addressed initial feedback. Removed unrelated code.

**Comment by alpaix** [comment]
Started working on next milestone. Closing the PR for now. Won't merge.

**Pull request closed without merging**